include/ghex/glue/atlas/atlas_user_concepts.hpp

```diff
@@ -0,0 +1,473 @@
+         * An Atlas domain has already the notion of halos.
+         * The purpose of the halo generator is to fill a container
+         * in which each element (halo) gathers all the indexes
+         * referring to the same remote partition.
+         * @tparam DomainId domain id type*/
+        template<typename DomainId>
+        class atlas_halo_generator {
+
+            public:
+
+                // member types
+                using domain_type = atlas_domain_descriptor<DomainId>;
+                using index_t = atlas::idx_t;
+
+                /** @brief Halo concept for Atlas
+                 * Puts together a neighbouring partition index
+                 * and the sequences of indexes of the elements of that partition.
+                 * Conceptually it is similar to an iteration space
+                 * and the two concepts could potentially coincide.
+                 * Stores information on the number of vertical levels as well,
+                 * but its size corresponds only to the horizontal size.*/
+                class halo {
+
+                    private:
+
+                        int m_partition;
+                        std::vector<index_t, gridtools::ghex::allocator::cuda::unified_memory_allocator<index_t>> m_local_index;
+                        std::vector<index_t> m_remote_index;
+                        std::size_t m_levels;
+
+                    public:
+
+                        // ctors
+                        halo() noexcept = default;
+                        halo(const int partition, const std::size_t levels = 1) noexcept :
+                            m_partition{partition},
+                            m_local_index{},
+                            m_remote_index{},
+                            m_levels{levels} {}
+                        // following one not strictly needed
+                        halo(const int partition,
+                             const std::vector<index_t, gridtools::ghex::allocator::cuda::unified_memory_allocator<index_t>>& local_index,
+                             const std::vector<index_t>& remote_index,
+                             const std::size_t levels = 1) noexcept :
+                            m_partition{partition},
+                            m_local_index{local_index},
+                            m_remote_index{remote_index},
+                            m_levels{levels} {
+                            assert(local_index.size() == remote_index.size());
+                        }
+
+                        // member functions
+                        int partition() const noexcept { return m_partition; }
+                        std::vector<index_t, gridtools::ghex::allocator::cuda::unified_memory_allocator<index_t>>& local_index() noexcept { return m_local_index; }
+                        const std::vector<index_t, gridtools::ghex::allocator::cuda::unified_memory_allocator<index_t>>& local_index() const noexcept { return m_local_index; }
```

Comment:
wouldn't it make sense to have a typedef for all these long templated types?

Reply:
Sure, done.